lib/msal-browser/src/interaction_client/PopupClient.ts

```diff
@@ -250,4 +259,274 @@ export class PopupClient extends StandardInteractionClient {
+        } else {
+            // Throw error if request URL is empty.
+            this.logger.error("Navigate url is empty");
+            throw BrowserAuthError.createEmptyNavigationUriError();
+        }
+    }
+
+    /**
+     * Monitors a window until it loads a url with the same origin.
+     * @param popupWindow - window that is being monitored
+     * @param timeout - timeout for processing hash once popup is redirected back to application
+     */
+    monitorPopupForHash(popupWindow: Window): Promise<string> {
+        return new Promise((resolve, reject) => {
+            /*
+             * Polling for popups needs to be tick-based,
+             * since a non-trivial amount of time can be spent on interaction (which should not count against the timeout).
+             */
+            const maxTicks = this.config.system.windowHashTimeout / BrowserConstants.POLL_INTERVAL_MS;
+            let ticks = 0;
+
+            this.logger.verbose("PopupHandler.monitorPopupForHash - polling started");
+
+            const intervalId = setInterval(() => {
+                // Window is closed
+                if (popupWindow.closed) {
+                    this.logger.error("PopupHandler.monitorPopupForHash - window closed");
+                    this.cleanPopup();
+                    clearInterval(intervalId);
+                    reject(BrowserAuthError.createUserCancelledError());
+                    return;
+                }
+
+                let href: string = Constants.EMPTY_STRING;
+                let hash: string = Constants.EMPTY_STRING;
+                try {
+                    /*
+                     * Will throw if cross origin,
+                     * which should be caught and ignored
+                     * since we need the interval to keep running while on STS UI.
+                     */
+                    href = popupWindow.location.href;
+                    hash = popupWindow.location.hash;
+                } catch (e) {}
+                
+                // Don't process blank pages or cross domain
+                if (StringUtils.isEmpty(href) || href === "about:blank") {
+                    return;
+                }
+
+                this.logger.verbose("PopupHandler.monitorPopupForHash - popup window is on same origin as caller");
+
+                /*
+                 * Only run clock when we are on same domain for popups
+                 * as popup operations can take a long time.
+                 */
+                ticks++;
+
+                if (hash) {
+                    this.logger.verbose("PopupHandler.monitorPopupForHash - found hash in url");
+                    clearInterval(intervalId);
+                    this.cleanPopup(popupWindow);
+    
+                    if (UrlString.hashContainsKnownProperties(hash)) {
+                        resolve(hash);
+                    } else {
+                        reject(BrowserAuthError.createHashDoesNotContainKnownPropertiesError());
```

Comment:
Can we add an error log message here with notes about an app router potentially clearing the hash (in addition the text in the error itself)?

Reply:
Done. Also added these errors to the error docs and included a link to the docs in the error message.